src/Distribution/Nixpkgs/Nodejs/License.hs

```diff
@@ -0,0 +1,128 @@
+{-# LANGUAGE NoImplicitPrelude, OverloadedStrings, RecordWildCards, FlexibleInstances, GeneralizedNewtypeDeriving #-}
+{-|
+Description: Convert @package.json@ license fields to nixpkgs license attribute sets
+-}
+module Distribution.Nixpkgs.Nodejs.License
+  ( -- * Conversion Logic
+    nodeLicenseToNixpkgs
+    -- * License Set Representation
+  , NixpkgsLicense (..)
+  , unfreeLicense
+    -- * License Lookup Table
+  , decode
+  , LicensesBySpdxId (..)
+  , lookupSpdxId
+  ) where
+
+import Protolude
+
+import Data.Aeson ((.:), (.:?))
+import qualified Data.Aeson as A
+import qualified Data.Aeson.Types as AT
+import qualified Data.ByteString.Lazy as BL
+import qualified Data.HashMap.Strict as HMS
+import qualified Data.HashMap.Lazy as HML
+import Nix.Expr
+import Distribution.Nixpkgs.Nodejs.Utils (attrSetMay, attrSetMayStr)
+
+-- newtype to circumvent the default instance: we don't want
+-- the key of the JSON object to be the key of the HashMap,
+-- but one of its values (spdxId).
+-- | Lookup table from SPDX identifier (as 'Text') to 'NixpkgsLicense'.
+--
+--   Should be obtained using 'decode'.
+newtype LicensesBySpdxId
+  = LicensesBySpdxId { unLicensesBySpdxId :: HML.HashMap Text NixpkgsLicense }
+  deriving (Show, Eq, Semigroup, Monoid)
+
+-- | Representation of a nixpkgs license set as found in
+--   @lib.licenses@. There doesn't seem to be a strict
+--   definition of what is required and what is optional,
+--   the distribution of 'Maybe' and non-'Maybe' values
+--   is based on the current situation in @lib/licenses.nix@.
+data NixpkgsLicense
+  = NixpkgsLicense
+  { attrName  :: Text       -- ^ Attribute name of the license in @lib.licenses@
+  , shortName :: Text
+  , spdxId    :: Maybe Text
+  , fullName  :: Text
+  , url       :: Maybe Text
+  , free      :: Maybe Bool
+  } deriving (Show, Eq)
+
+-- | Static version of @lib.licenses.unfree@,
+--   so @UNLICENSED@ can be handled correctly
+--   even if no lookup table is provided.
+unfreeLicense :: NixpkgsLicense
+unfreeLicense = NixpkgsLicense
+  { attrName = "unfree"
+  , shortName = "unfree"
+  , fullName = "Unfree"
+  , free = Just False
+  , spdxId = Nothing
+  , url = Nothing
+  }
+
+instance A.FromJSON LicensesBySpdxId where
+  parseJSON = A.withObject "NixpkgsLicenseSet" $
+    HMS.foldrWithKey (\k v p -> addNixpkgsLicense k v =<< p) (pure mempty)
+
+addNixpkgsLicense :: Text -> AT.Value -> LicensesBySpdxId -> AT.Parser LicensesBySpdxId
+addNixpkgsLicense attr val lics = do
+  license <- A.withObject "NixpkgsLicense" parseLicense val
+  -- insert if it has an spdxId, otherwise just return lics
+  pure . fromMaybe lics $ LicensesBySpdxId
+    . flip (flip HML.insert license) (unLicensesBySpdxId lics)
+    <$> spdxId license
```

Comment:
```
let (LicensesBySpdxId licsMap) = lics
case license & spdxId of
  Nothing -> pure lics
  Just i -> pure $ LicensesBySpdxId $ HML.insert i license licsMap
  ```
  
  lol, that took a while to untangle

Reply:
As a rule of thumb, once you use `flip` you are doing something wrong. Same with using `.` for constructor application. Same with using `<$>` in a context different from the outer `Monad` (i.e. the `do` block here is a `AT.Parser` but the `<$>` here is mapping over a `Maybe`.

Reply:
Done in ee7f26507bf12993df9c49fd88144261f8b355c2